pss/trojan_chunk.go

```diff
@@ -0,0 +1,208 @@
+
+	// set random bytes as padding
+	paddingLength := trojanPayloadMaxSize - payloadLength
+	padding := make([]byte, paddingLength)
+	if _, err := rand.Read(padding); err != nil {
+		return trojanMessage{}, err
+	}
+
+	// create new trojan message var and set fields
+	tm := new(trojanMessage)
+	copy(tm.length[:], lengthBuffer[:])
+	tm.payload = payload
+	tm.padding = padding
+
+	return *tm, nil
+}
+
+// newTrojanChunk creates a new trojan chunk for the given targets and trojan message
+// TODO: discuss if instead of receiving a trojan message, we should receive a byte slice as payload
+func newTrojanChunk(targets [][]byte, message trojanMessage) (chunk.Chunk, error) {
+	if err := checkTargets(targets); err != nil {
+		return nil, err
+	}
+
+	// create span
+	span := newTrojanChunkSpan()
+
+	// iterate trojan chunk fields to find coherent address and payload for chunk
+	addr, payload, err := iterateTrojanChunk(targets, span, message)
+	if err != nil {
+		return nil, err
+	}
+
+	return chunk.NewChunk(addr, payload), nil
+}
+
+// checkTargets verifies that the list of given targets is non empty and with elements of matching size
+func checkTargets(targets [][]byte) error {
+	if len(targets) == 0 {
+		return fmt.Errorf("target list cannot be empty")
+	}
+	validLength := len(targets[0]) // take first element as allowed length
+	for i := 1; i < len(targets); i++ {
+		if len(targets[i]) != validLength {
+			return fmt.Errorf("target list cannot have targets of different length")
+		}
+	}
+	return nil
+}
+
+// newTrojanChunkSpan creates a pre-set 8-byte span for a trojan chunk
+func newTrojanChunkSpan() []byte {
+	span := make([]byte, 8)
+	binary.BigEndian.PutUint64(span, chunk.DefaultSize) // TODO: should this be little-endian?
+	return span
+}
+
+// iterateTrojanChunk finds a nonce so that when the given trojan chunk fields are hashed, the result will fall in the neighbourhood of one of the given targets
+// this is done by iterating the BMT hash of the serialization of the trojan chunk fields until the desired nonce is found
+// the function returns the matching hash to be used as an address for a regular chunk, plus its payload
+// the payload is the serialization of the trojan chunk fields which correctly hash into the matching address
+func iterateTrojanChunk(targets [][]byte, span []byte, message trojanMessage) (addr, payload []byte, err error) {
+	emptyAddr, emptyPayload := []byte{}, []byte{}
+
+	// start out with random nonce
+	nonce := make([]byte, 32)
+	if _, errRand := rand.Read(nonce); err != nil {
+		return emptyAddr, emptyPayload, errRand
+	}
+	nonceInt := new(big.Int).SetBytes(nonce)
+	targetsLength := len(targets[0])
+
+	// serialize trojan message
+	m, marshalErr := message.MarshalBinary() // TODO: this should be encrypted
+	if marshalErr != nil {
+		return emptyAddr, emptyPayload, marshalErr
+	}
+
+	// hash trojan chunk fields with different nonces until an acceptable one is found
+	// TODO: prevent infinite loop
+	for {
```

Comment:
so here we can do parallelisation and also based on collision resistance, just make sure to generate different hashes.

```golang

seeds := [8]uint32
for i, _ := range seeds {
   seeds[i] = rand.Uint32()
}
nonce := make([]byte, 32)
for i:=0; i<32; i+=4 {
   binary.PutUint32(nonce[i:i+4], seeds[i])
}
for i:=0; i<8; i++ {
   seed :=  make([]byte,32)
   copy(seed, nonce)
   go hashWorker(seed, payload, quitc, resultc)
}

func hashWorker()
```